dace/codegen/targets/fpga_helper/fpga_utils.py

```diff
@@ -0,0 +1,194 @@
+# Copyright 2019-2021 ETH Zurich and the DaCe authors. All rights reserved.
+import copy
+from typing import List, Union, Tuple
+from dace import data as dt, SDFG, dtypes, subsets, symbolic
+
+_FPGA_STORAGE_TYPES = {
+    dtypes.StorageType.FPGA_Global, dtypes.StorageType.FPGA_Local,
+    dtypes.StorageType.FPGA_Registers, dtypes.StorageType.FPGA_ShiftRegister
+}
+
+
+def is_hbm_array(array: dt.Data):
+    """
+    :return: True if this array is placed on HBM
+    """
+    if (isinstance(array, dt.Array)
+            and array.storage == dtypes.StorageType.FPGA_Global):
+        res = parse_location_bank(array)
+        return res is not None and res[0] == "HBM"
+    else:
+        return False
+
+
+def is_fpga_array(array: dt.Data):
+    """
+    :return: True if this array is placed on FPGA memory
+    """
+    return isinstance(array, dt.Array) and array.storage in _FPGA_STORAGE_TYPES
+
+
+def iterate_hbm_multibank_arrays(array_name: str, array: dt.Array, sdfg: SDFG):
+    """
+    Small helper function that iterates over the bank indices
+    if the provided array is spanned across multiple HBM banks.
+    Otherwise just returns 0 once.
+    """
+    res = parse_location_bank(array)
+    if res is not None:
+        bank_type, bank_place = res
+        if (bank_type == "HBM"):
+            low, high = get_multibank_ranges_from_subset(bank_place, sdfg)
+            for i in range(high - low):
+                yield i
+        else:
+            yield 0
+    else:
+        yield 0
+
+
+def modify_distributed_subset(subset: Union[subsets.Subset, list, tuple],
+                              change: int):
+    """
+    Modifies the first index of :param subset: (the one used for distributed subsets).
+    :param subset: is deepcopied before any modification to it is done.
+    :param change: the first index is set to this value, unless it's (-1) in which case
+        the first index is completly removed
+    """
+    cps = copy.deepcopy(subset)
+    if isinstance(subset, subsets.Subset):
+        if change == -1:
+            cps.pop([0])
+        else:
+            cps[0] = (change, change, 1)
+    elif isinstance(subset, list) or isinstance(subset, tuple):
+        if isinstance(subset, tuple):
+            cps = list(cps)
+        if change == -1:
+            cps.pop(0)
+        else:
+            cps[0] = change
+        if isinstance(subset, tuple):
+            cps = tuple(cps)
```

Comment:
Couldn't this be anything iterable? I think you can replace all of this with just:
```python
cps = list(cps)  # Works for any iterable type
if change == -1:
    cps = cps[1:]
else:
    cps[0] = change
```
The input type would then just be `Iterable`.